### PR TITLE
[IntelliJ Plugin] Improve source root detection when running ballerina files 

### LINKED ISSUE #10369: [IDEA] Getting ballerina project base path instead of IDEA project base path as the source root
**Description:**
When we currently run/debug a ballerina program, we get the IDEA project base path as the source root. Instead of that, we need to get the root of ballerina project.

This change needs to be done when we set the `--sourceroot` parameter.

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose
Resolves #10369 and #10093

### CODE REVIEW

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java [hunk 1]
```diff
@@ -67,18 +68,36 @@ public void startNotified(ProcessEvent event) {
...
+        String filePath;
 
         // Find the file in the project. This is needed to find the module. Otherwise if the file is in a sub-module
         // and the SDK for the project is not set, SDK home path will be null.
         PsiFile file = BallerinaPsiImplUtil.findFileInProject(project, myConfiguration.getFilePath());
-        Module module = null;
-        if (file != null) {
-            module = ModuleUtilCore.findModuleForPsiElement(file);
+
+        assert file != null;
```
Comment (Shan1024) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Remove `assert` and use an `if` condition.

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java [hunk 2]
```diff
@@ -103,22 +117,24 @@ protected BallerinaExecutor patchExecutor(@NotNull BallerinaExecutor executor) t
...
+            ballerinaExecutor.withBallerinaPath(
                     BallerinaSdkService.getInstance(getConfiguration().getProject()).getSdkHomePath(module))
-                    .withParameterString(myConfiguration.getBallerinaToolParams()).withParameters(filePath);
+                    .withParameterString(myConfiguration.getBallerinaToolParams());
         }
 
+        // Adds ballerina file/package name after flags.
+        ballerinaExecutor.withParameters(filePath);
+
         return ballerinaExecutor;
```
Comment (Shan1024) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Check whether we can merge these last 2 lines.

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java [hunk 3]
```diff
@@ -138,4 +154,28 @@ public void setHistoryProcessHandler(@Nullable BallerinaHistoryProcessListener h
...
+     * directory is found.
+     */
+    private String getSourceRoot(String currentPath, String root) {
+
+        if (currentPath.equals(root) || currentPath.equals("") || root.equals("")) {
+            return "";
+        } else {
+            File currentDir = new File(currentPath);
+            File[] files = currentDir.listFiles();
+            if (files != null) {
```
Comment (Shan1024) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Invert the `if` condition.

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java [hunk 4]
```diff
@@ -67,18 +68,36 @@ public void startNotified(ProcessEvent event) {
...
+        // Sets source root of the IDEA project as the termination of the recursive search.
+        String rootDir = project.getBasePath();
+        String sourcerootDir = getSourceRoot(fileDir.getPath(), rootDir);
+
+        if (sourcerootDir.equals("")) {  // if no ballerina project is found.
+            sourcerootDir = baseDir.getPath();
+            filePath = file.getVirtualFile().getPath().replace(sourcerootDir, "").substring(1);
+        } else {
+            String relativeFilePath = file.getVirtualFile().getPath().replace(sourcerootDir, "").substring(1);
+            if (relativeFilePath.contains(File.separator)) { //If file is found in a ballerina package, runs the whole
```
Comment (Shan1024) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Move the comment above the `if` statement.

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java [hunk 5]
```diff
@@ -67,18 +68,36 @@ public void startNotified(ProcessEvent event) {
 
     @Override
     protected BallerinaExecutor patchExecutor(@NotNull BallerinaExecutor executor) throws ExecutionException {
-        RunConfigurationKind kind = getConfiguration().getRunKind();
+        // RunConfigurationKind kind = getConfiguration().getRunKind();
```
Comment (Shan1024) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Lets remove the commented codes as well.

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java [hunk 6]
```diff
@@ -67,18 +68,36 @@ public void startNotified(ProcessEvent event) {
...
+            if (relativeFilePath.contains(File.separator)) { //If file is found in a ballerina package, runs the whole
+                // package, instead of the single file.
+                filePath = relativeFilePath.substring(0, relativeFilePath.indexOf(File.separator));
+            } else {
+                filePath = relativeFilePath;
+            }
         }
+
+        Module module;
+        module = ModuleUtilCore.findModuleForPsiElement(file);
```
Comment (anupama-pathirage) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Lets merge the two lines. No need to have variable `module` declaration separately

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Fixed

Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Done

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java (hunk 3 above)
Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Done

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java (hunk 4 above)
Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Done

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java (hunk 5 above)
Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Removed

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java (hunk 6 above)
Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/runconfig/application/BallerinaApplicationRunningState.java]: Merged

**Pull request merged**